include/swift/AST/Decl.h

```diff
@@ -807,6 +800,15 @@ class alignas(1 << DeclAlignInBits) Decl {
   /// in diagnostics.
   SourceLoc getLoc(bool SerializedOK = true) const;
 
+  /// Returns the serialized locations of this declaration from the
+  /// corresponding .swiftsourceinfo file. Empty if it is within the current
+  /// module and thus does not have any .swiftsourceinfo.
+  ///
+  /// By default the locations are simple file, offset, line, and column. To
+  /// resolve these into \c SourceLocs, set \p Resolve to \c true. Note that
```

Comment:
I'll expand the comment.

`Empty` really means "default", ie. `Positions` is a nullptr, `BufferID` is 0, `Location` is invalid, etc. Resolve false vs true: `Positions` set either way (unless no .swiftsourceinfo), resolve sets BufferId/Loc/DocRanges.

An alternative here would be to have two methods instead - one that returns BasicDeclPositions and the other ExternalDeclLocs. Then remove Positions and Resolved from ExternalDeclLocs. The cache would be a std::pair of BasicDeclPositions and ExternalDeclLocs. Only downside there is that it'd be weird to then have BasicDeclPositions as a nested type in ExternalDeclLocs so we'd really need a new name.

Reply:
> Empty really means "default", ie. Positions is a nullptr, BufferID is 0, Location is invalid, etc. Resolve false vs true: Positions set either way (unless no .swiftsourceinfo), resolve sets BufferId/Loc/DocRanges.

👍 

> An alternative here would be to have two methods instead 

I think the current approach is probably better for clarity, on balance.  It just needs a bit more verbiage about how the pieces come together.  It might make sense to move/duplicate some of this information in doc comments for `ExternalDeclLocs` itself, or on its individual fields.  It's a bit of a tricky type, so being a bit verbose about it makes sense to me.